linkerd/stack/src/proxy.rs

```diff
@@ -0,0 +1,87 @@
+use futures::{future, Future, Poll};
+use linkerd2_error::Error;
+
+/// A middleware type that cannot exert backpressure.
+///
+/// Typically used to modify requests or responses.
+pub trait Proxy<Req, S: tower::Service<Self::Request>> {
+    /// The type of request sent to the inner `S`-typed service.
+    type Request;
+
+    /// The type of response returned to callers.
+    type Response;
+
+    /// The error type returned to callers.
+    type Error: Into<Error>;
+
+    /// The Future type returned to callers.
+    type Future: Future<Item = Self::Response, Error = Self::Error>;
+
+    /// Usually invokes `S::call`, potentially modifying requests or responses.
+    fn proxy(&self, inner: &mut S, req: Req) -> Self::Future;
+
+    /// Wraps an `S` typed service with the proxy.
+    fn into_service(self, inner: S) -> ProxyService<Self, S>
```

Comment:
also admittedly very bikesheddy, but I kind of expect `into_` methods to be conversions that only take `self`. i might've called this `wrap` or `and_then` or something...